File: python/pollax/resources/campaigns.py
"""Campaigns resource"""

from typing import List, Optional, TYPE_CHECKING
from ..models import Campaign, Contact

if TYPE_CHECKING:
    from ..client import Pollax


class CampaignsResource:
    """Campaigns API resource"""

    def __init__(self, client: "Pollax"):
        self._client = client

    def create(
        self,
        name: str,
        agent_id: Optional[str] = None,
        scheduled_time: Optional[str] = None,
        contacts: Optional[List[dict]] = None,
        enable_voicemail: Optional[bool] = None,
        voicemail_message: Optional[str] = None,
        announcement_message: Optional[str] = None,
        *,
        idempotency_key: Optional[str] = None,
    ) -> Campaign:
        """Create a new campaign.

        Args:
            name: Campaign display name.
            agent_id: Agent to dispatch on every call.
            scheduled_time: Optional ISO 8601 timestamp to start the campaign.
            contacts: List of contact dicts (phone, name, metadata).
            idempotency_key: Optional ``Idempotency-Key`` header value.
                A retry with the same key and same body returns the cached
                response — protects against double-creating a campaign on
                client retry.
        """
        data = {
            "name": name,
            "agent_id": agent_id,
            "scheduled_time": scheduled_time,
            "contacts": contacts or [],
        }
        if enable_voicemail is not None:
            data["enable_voicemail"] = enable_voicemail
        if voicemail_message:
            data["voicemail_message"] = voicemail_message
        if announcement_message:
            data["announcement_message"] = announcement_message
        headers = {"Idempotency-Key": idempotency_key} if idempotency_key else None
        response = self._client.request("POST", "/api/v1/campaigns", json=data, headers=headers)
        return Campaign(**response)

    def list(self) -> List[Campaign]:
        """List all campaigns."""
        response = self._client.request("GET", "/api/v1/campaigns")
        return [Campaign(**item) for item in response]

    def retrieve(self, campaign_id: str) -> Campaign:
        """Get a single campaign by ID."""
        response = self._client.request("GET", f"/api/v1/campaigns/{campaign_id}")
        return Campaign(**response)

    def update(self, campaign_id: str, **kwargs) -> Campaign:
        """Update a campaign."""
        response = self._client.request("PUT", f"/api/v1/campaigns/{campaign_id}", json=kwargs)
        return Campaign(**response)

    def delete(self, campaign_id: str) -> dict:
        """Delete a campaign."""
        return self._client.request("DELETE", f"/api/v1/campaigns/{campaign_id}")

    def start(self, campaign_id: str) -> Campaign:
        """Start a campaign."""
        response = self._client.request("POST", f"/api/v1/campaigns/{campaign_id}/start")
        return Campaign(**response)

    def pause(self, campaign_id: str) -> Campaign:
        """Pause a running campaign."""
        response = self._client.request("POST", f"/api/v1/campaigns/{campaign_id}/pause")
        return Campaign(**response)

    def get_stats(self, campaign_id: str) -> dict:
        """Get campaign statistics."""
        return self._client.request("GET", f"/api/v1/campaigns/{campaign_id}/stats")

File: typescript/src/resources/campaigns.ts
import { AxiosRequestConfig } from 'axios';
import {
  Campaign,
  CreateCampaignParams,
  UpdateCampaignParams,
} from '../types';
import { RequestOptions } from './calls';

function buildHeaders(opts?: RequestOptions): Record<string, string> | undefined {
  if (opts?.idempotencyKey) {
    return { 'Idempotency-Key': opts.idempotencyKey };
  }
  return undefined;
}

export class Campaigns {
  constructor(private request: <T = any>(config: AxiosRequestConfig) => Promise<T>) {}

  /**
   * Create a new campaign.
   *
   * @example
   * const campaign = await pollax.campaigns.create(
   *   {
   *     name: 'Q1 Outreach Campaign',
   *     agent_id: 'agent_123',
   *     contacts: [
   *       { name: 'John Doe', phone: '+1234567890' },
   *       { name: 'Jane Smith', phone: '+0987654321' },
   *     ],
   *   },
   *   { idempotencyKey: 'q1-outreach-launch' }
   * );
   */
  async create(params: CreateCampaignParams, options?: RequestOptions): Promise<Campaign> {
    return this.request<Campaign>({
      method: 'POST',
      url: '/api/v1/campaigns',
      data: params,
      headers: buildHeaders(options),
    });
  }

  /**
   * List all campaigns
   * 
   * @example
   * const campaigns = await pollax.campaigns.list();
   */
  async list(): Promise<Campaign[]> {
    return this.request<Campaign[]>({
      method: 'GET',
      url: '/api/v1/campaigns',
    });
  }

  /**
   * Get a single campaign by ID
   * 
   * @example
   * const campaign = await pollax.campaigns.retrieve('campaign_123');
   */
  async retrieve(campaignId: string): Promise<Campaign> {
    return this.request<Campaign>({
      method: 'GET',
      url: `/api/v1/campaigns/${campaignId}`,
    });
  }

  /**
   * Update a campaign
   * 
   * @example
   * const campaign = await pollax.campaigns.update('campaign_123', {
   *   status: 'paused',
   * });
   */
  async update(campaignId: string, params: UpdateCampaignParams): Promise<Campaign> {
    return this.request<Campaign>({
      method: 'PUT',
      url: `/api/v1/campaigns/${campaignId}`,
      data: params,
    });
  }

  /**
   * Delete a campaign
   * 
   * @example
   * await pollax.campaigns.delete('campaign_123');
   */
  async delete(campaignId: string): Promise<{ success: boolean }> {
    return this.request({
      method: 'DELETE',
      url: `/api/v1/campaigns/${campaignId}`,
    });
  }

  /**
   * Start a campaign
   * 
   * @example
   * await pollax.campaigns.start('campaign_123');
   */
  async start(campaignId: string): Promise<Campaign> {
    return this.request<Campaign>({
      method: 'POST',
      url: `/api/v1/campaigns/${campaignId}/start`,
    });
  }

  /**
   * Pause a running campaign
   * 
   * @example
   * await pollax.campaigns.pause('campaign_123');
   */
  async pause(campaignId: string): Promise<Campaign> {
    return this.request<Campaign>({
      method: 'POST',
      url: `/api/v1/campaigns/${campaignId}/pause`,
    });
  }

  /**
   * Resume a paused campaign.
   *
   * @example
   * await pollax.campaigns.resume('campaign_123');
   */
  async resume(campaignId: string): Promise<Campaign> {
    return this.request<Campaign>({
      method: 'POST',
      url: `/api/v1/campaigns/${campaignId}/resume`,
    });
  }

  /**
   * Stop a campaign (mark it complete; no further calls are placed).
   *
   * @example
   * await pollax.campaigns.complete('campaign_123');
   */
  async complete(campaignId: string): Promise<Campaign> {
    return this.request<Campaign>({
      method: 'POST',
      url: `/api/v1/campaigns/${campaignId}/complete`,
    });
  }

  /**
   * Upload contacts to a campaign via CSV
   * 
   * @example
   * const campaign = await pollax.campaigns.uploadContacts('campaign_123', csvFile);
   */
  async uploadContacts(campaignId: string, file: File | Buffer): Promise<Campaign> {
    const formData = new FormData();
    formData.append('file', file);

    return this.request<Campaign>({
      method: 'POST',
      url: `/api/v1/campaigns/${campaignId}/upload`,
      data: formData,
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
  }

  /**
   * Get campaign statistics
   * 
   * @example
   * const stats = await pollax.campaigns.getStats('campaign_123');
   */
  async getStats(campaignId: string): Promise<{
    total_contacts: number;
    completed_calls: number;
    successful_calls: number;
    failed_calls: number;
    success_rate: string;
  }> {
    return this.request({
      method: 'GET',
      url: `/api/v1/campaigns/${campaignId}/stats`,
    });
  }
}
